File: example-android-project/app/src/main/java/io/github/gofaith/faithdroid/FViews/FConstraintLayout.java
package io.github.gofaith.faithdroid.FViews;

import android.support.constraint.ConstraintLayout;
import android.util.Log;

import java.util.Map;

import io.github.gofaith.faithdroid.UI.AttrGettable;
import io.github.gofaith.faithdroid.UI.AttrSettable;
import io.github.gofaith.faithdroid.UI.UIController;

public class FConstraintLayout extends FView implements AttrSettable,AttrGettable {
    public ConstraintLayout v;
    public FConstraintLayout(UIController controller) {
        parentController = controller;
        v = new ConstraintLayout(parentController.activity);
        view = v;
    }
    @Override
    public String getAttr(String attr) {
        String str = getUniversalAttr(attr);
        if (str != null) {
            return str;
        }
        switch (attr) {

        }
        return null;
    }

    @Override
    public void setAttr(String attr, String value) {
        if (setUniversalAttr(attr, value)) {
            return;
        }
        switch (attr) {
            case "Append":
                String[] vids = value.split(",");
                for (int i=0;i<vids.length;i++) {
                    FView f= parentController.viewmap.get(vids[i]);
                    ConstraintLayout.LayoutParams lp = new ConstraintLayout.LayoutParams(f.size[0], f.size[1]);
                    for (Map.Entry<String, ConstraintInterface> entry : f.afterConstraintFuncs.entrySet()) {
                        entry.getValue().addConstraint(this,lp);
                    }
                    v.addView(parentController.viewmap.get(vids[i]).view,lp);
                }
                break;
        }
    }
}

File: android/app/src/main/java/io/github/stevenzack/faithdroidapp/FViews/FConstraintLayout.kt
package io.github.stevenzack.faithdroidapp.FViews

import androidx.constraintlayout.widget.ConstraintLayout
import io.github.stevenzack.faithdroidapp.UI.AttrGettable
import io.github.stevenzack.faithdroidapp.UI.AttrSettable
import io.github.stevenzack.faithdroidapp.UI.FView
import io.github.stevenzack.faithdroidapp.UI.UIController


class FConstraintLayout(controller: UIController) : FView(), AttrSettable, AttrGettable {
    var v: ConstraintLayout

    init {
        parentController = controller
        v = ConstraintLayout(parentController!!.activity)
        view = v
    }

   override fun getAttr(attr: String): String {
        val str = getUniversalAttr(attr)
        if (str != null) {
            return str
        }
        when (attr) {

        }
        return ""
    }

   override fun setAttr(attr: String, value: String) {
        if (setUniversalAttr(attr, value)) {
            return
        }
        when (attr) {
            "Append" -> {
                val vids = value.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                for (i in vids.indices) {
                    val f = parentController!!.viewmap[vids[i]]
                    val lp = ConstraintLayout.LayoutParams(f!!.size[0], f!!.size[1])
                    for (entry in f.afterConstraintFuncs) {
                        entry.value.addConstraint(this, lp)
                    }
                    lp.leftMargin = f.margin[0]
                    lp.topMargin = f.margin[1]
                    lp.rightMargin = f.margin[2]
                    lp.bottomMargin = f.margin[3]
                    v.addView(parentController!!.viewmap[vids[i]]!!.view, lp)
                }
            }
        }
    }
}
